Use a lock to prevent pause/resume calls during polling

waj opened the pull request:
This adds a lock to prevent calls to pause/resume the consumer concurrently with the polling. There's a [bug](https://github.com/confluentinc/librdkafka/blob/c282ba2423b2694052393c8edb0399a5ef471b3f/CHANGELOG.md?plain=1#L90-L95) in the librdkafka library that corrupts the current offset stored in memory making it miss or process messages out of order.

Initially I thought about using an `MVar KafkaConsumer` but it's actually not required to enforce the serialization with other calls like `seek` and `assignment`.

It would be nice to add specs but I couldn't find an easy and fast enough way to reproduce the issue. To test this I was using a process that enqueues 300 messages and a process that just sleeps for 200ms on each message. That consistently made it skip messages once the buffer gets full.

Closes ZEN-984

stoeffel commented on nri-prelude/src/Examples.hs:
```diff
@@ -6,33 +6,37 @@
 module Examples
   ( HasExamples (..),
-    Example (..),
+    Examples,
```
We need `encodedValue` in `TestEncoding` though. https://github.com/NoRedInk/NoRedInk/blob/0fe4ff37327c1cf7ea5b2005c1ab9e39e7b02581/lib/webserver/src/Nri/WebServer/TestEncoding.hs#L54

I was thinking that we could opensource that as well and move `Examples` to `Examples.Internal`.

jwoudenberg-nri replied:
I was thinking maybe we can adjust our tests and create one test per type, using the `render` function to generate all we need?

jwoudenberg-nri replied:
Alternatively: Maybe we should open source a separate `nri-examples` library and have it expose all examples related helpers:
- Code for creating examples for types
- Code for using examples in combination with Servant
- Code for using examples in combination with Redis

Thoughts?

stoeffel replied:
I don't have a strong preference about separating into a different package. One test per type sounds good to me.

jwoudenberg-nri replied:
Context: I only think we should do the separate package approach if it allows us to avoid pulling a servant dependency into `nri-prelude`.

stoeffel replied:
I just added the necessary orphan instances to the `TestApiEncoding` module in `Nri.WebServer`. Do you think that's okay for now?

jwoudenberg-nri replied:
Ohhh, didn't consider there'd be orphans.

I think it's fine for now, but also that if we can avoid orphans by doing a separate lib it's worth doing, and then we might as well do it now.

jwoudenberg-nri replied:
Also in Zoom if we want to chat sync.

stoeffel commented on nri-prelude/src/Examples.hs:
```diff
@@ -41,19 +45,31 @@ example description x =
-
-toList :: (a, List a) -> List a
-toList (x, xs) = x : xs
+concat :: Examples -> Examples -> Examples
```
Maybe we should just make `Examples` a semigroup. Thoughts?

jwoudenberg-nri replied:
Also works for me!

jwoudenberg-nri replied:
Works for me too!

stoeffel commented:
merging this into the other branch

waj committed:
Use a lock to prevent pause/resume calls during polling

omnibs committed:
Document the bug we're working around

omnibs reviewed (commented):
I'm pondering whether it'd be nice for future us to leave your test producer/consumer around the codebase, to make it easy to check this bug is fixed, the day hw-kafka gets bumped to librdkafka >= 2.1.0.

We could:
- Add a `nri-kafka/script/pause-resume-bug/{producer,consumer}.hs`
- Add a readme there too
- Point to it from the comments

wdyt?

waj commented:
@omnibs that makes sense, I'll see if I can quickly convert my test into something worth to share.

waj committed:
Sample code to reproduce the pause/resume bug

omnibs committed:
Print all states

omnibs committed:
Make the consumer print all states, atomically

omnibs committed:
Use a partition key

There's no ordering guarantees otherwise, and it'll make the consumer
think it is reproducing the bug, when it isn't.

omnibs committed:
Clear expectations

omnibs committed:
Teach how to reproduce the bug

omnibs reviewed (approved):
Thank you so much for digging in, fixing, and providing a way to reproduce the bug!!!
